### PR TITLE
Fix possible memory corruption after realloc and memory leaks

### PR DESCRIPTION
**Pull request by abbat**
realloc can move memory block, so if we do not store pointer to it (if next malloc/realloc failed) we can get pointer to invalid memory block and crash application on next access.

### CODE REVIEW

File: toxcore/group_chats.c [hunk 1]
```diff
@@ -4429,6 +4429,8 @@ static int peer_add(Messenger *m, int groupnumber, IP_Port *ipp, const uint8_t *
             return -1;
     }
 
+    chat->gcc = tmp_gcc;
```
Comment (JFreegman) [inline toxcore/group_chats.c]: This makes no sense

File: toxcore/group_moderation.c [hunk 2]
```diff
@@ -60,8 +60,15 @@ int mod_list_unpack(GC_Chat *chat, const uint8_t *data, uint32_t length, uint32_
     for (i = 0; i < num_mods; ++i) {
         tmp_list[i] = malloc(sizeof(uint8_t) * GC_MOD_LIST_ENTRY_SIZE);
 
-        if (tmp_list[i] == NULL)
+        if (tmp_list[i] == NULL) {
```
Comment (JFreegman) [inline toxcore/group_moderation.c]: This is ugly. Just add the line: `free_uint8_t_pointer_array(tmp_list, i);`

File: toxcore/group_moderation.c [hunk 3]
```diff
@@ -220,10 +227,10 @@ int mod_list_add_entry(GC_Chat *chat, const uint8_t *mod_data)
 
     uint8_t **tmp_list = realloc(chat->moderation.mod_list, sizeof(uint8_t *) * (chat->moderation.num_mods + 1));
 
-    if (tmp_list == NULL) {
-        chat->moderation.num_mods = 0;
+    if (tmp_list == NULL)
         return -1;
-    }
+
+    chat->moderation.mod_list = tmp_list;
```
Comment (JFreegman) [inline toxcore/group_moderation.c]: This is wrong

**Comment by JFreegman** [comment]
There are a few problems with this PR that I commented on. I fixed what needs to be fixed in a separate commit here: https://github.com/JFreegman/toxcore/commit/8fe0952ccce226cf5feb43ba67dfe1a68bf50d67

It's nice to have another set of eyes on the code. In the future just always remember to make sure your code compiles before making a pull request.

File: toxcore/group_chats.c (hunk 1 above)
Comment (abbat) [inline toxcore/group_chats.c]: Sorry, Friday night.

```
GC_Connection *tmp_gcc = realloc(chat->gcc, ...
if (tmp_gcc == NULL) { ... }

GC_GroupPeer *tmp_group = realloc...

if (tmp_group == NULL) {
    // after realloc fails chat->gcc may be invalid
```

Comment (abbat) [inline toxcore/group_chats.c]: chat->gcc = tmp_gcc; must be after 

```
    if (tmp_gcc == NULL) {
        kill_tcp_connection_to(chat->tcp_conn, tcp_connection_num);
        return -1;
    }
```

File: toxcore/group_moderation.c (hunk 2 above)
Comment (abbat) [inline toxcore/group_moderation.c]: Agree.

File: toxcore/group_moderation.c (hunk 3 above)
Comment (abbat) [inline toxcore/group_moderation.c]: No, I think. mod_list just resized, but not initialized and num_mods not chaged. Next call for realloc with same size is safe.

**Comment by abbat** [comment]
I think it is last PR. Next after merge. Thanks for your work!

File: toxcore/group_chats.c (hunk 1 above)
Comment (JFreegman) [inline toxcore/group_chats.c]: If realloc fails, the original pointer will not be touched.

Comment (abbat) [inline toxcore/group_chats.c]: There is 2 reallocs, one may be success.

Comment (JFreegman) [inline toxcore/group_chats.c]: I see now, fixed: https://github.com/JFreegman/toxcore/commit/bfbab7e620f101f2752e4b8acceb5ddfa7204792 It's worth pointing out that if realloc fails, recovery is unlikely.

**Comment by JFreegman** [comment]
I've addressed everything here so I'm going to close the pull request. Thanks!

**Pull request closed without merging**